test/modules/TestResultCoordinator/CountingReportGenerator.cs

```diff
@@ -0,0 +1,179 @@
+            int batchSize = 500)
+        {
+            this.expectedSource = Preconditions.CheckNonWhiteSpace(expectedSource, nameof(expectedSource));
+            this.expectedStore = Preconditions.CheckNotNull(expectedStore, nameof(expectedStore));
+            this.actualSource = Preconditions.CheckNonWhiteSpace(actualSource, nameof(actualSource));
+            this.actualStore = Preconditions.CheckNotNull(actualStore, nameof(actualStore));
+            this.resultType = Preconditions.CheckNonWhiteSpace(resultType, nameof(resultType));
+            this.testResultComparer = Preconditions.CheckNotNull(testResultComparer, nameof(testResultComparer));
+            this.batchSize = batchSize;
+        }
+
+        public string ActualSource => this.actualSource;
+
+        public ISequentialStore<TestOperationResult> AcutalStore => this.actualStore;
+
+        public string ExpectedSource => this.expectedSource;
+
+        public ISequentialStore<TestOperationResult> ExpectedStore => this.expectedStore;
+
+        public string ResultType => this.resultType;
+
+        public ITestResultComparer<TestOperationResult> TestResultComparer => this.testResultComparer;
+
+        public int BatchSize => this.batchSize;
+
+        public async Task<ITestResultReport> CreateReportAsync()
+        {
+            Logger.LogInformation($"Start to generate report by {nameof(CountingReportGenerator)} for Sources [{this.expectedSource}] and [{this.actualSource}]");
+
+            long lastLoadedKeyFromExpectStore = -1;
+            long lastLoadedKeyFromActualStore = -1;
+            TestOperationResult lastLoadedResult = default(TestOperationResult);
+
+            this.report = new CountingReport<TestOperationResult>(this.expectedSource, this.actualSource, this.resultType);
+            var expectQueue = new Queue<TestOperationResult>();
+            var actualQueue = new Queue<TestOperationResult>();
+
+            (lastLoadedKeyFromExpectStore, _) = await this.LoadBatchIntoQueueAsync(this.expectedSource, this.expectedStore, lastLoadedKeyFromExpectStore, expectQueue);
+            (lastLoadedKeyFromActualStore, lastLoadedResult) = await this.LoadBatchIntoQueueAsync(this.actualSource, this.actualStore, lastLoadedKeyFromActualStore, actualQueue, lastLoadedResult);
+
+            while (expectQueue.Count > 0 && actualQueue.Count > 0)
+            {
+                TestOperationResult expectedResult = expectQueue.Dequeue();
+                TestOperationResult actualResult = actualQueue.Peek();
+
+                this.report.TotalExpectCount++;
+
+                if (this.testResultComparer.Matches(expectedResult, actualResult))
+                {
+                    actualQueue.Dequeue();
+                    this.report.TotalMatchCount++;
+                }
+                else
+                {
+                    this.report.AddMissingResult(expectedResult);
+                }
+
+                if (expectQueue.Count == 0 || actualQueue.Count == 0)
```

Comment:
Should this be moved out of the while loop?

Reply:
This is needed to continue to load another batch for verification.